packages/utilities/src/asAsync.tsx

```diff
@@ -0,0 +1,166 @@
+import * as React from 'react';
+
+interface IAsAsyncState<TProps> {
+  component?: React.ComponentType<TProps> | undefined;
+  error?: Error;
+  placeholder?: JSX.Element;
+  onLoad?: () => void;
+  onError?: (error: Error) => void;
+}
+
+/**
+ * If possible, use a WeakMap to maintain a cache of loaded components.
+ * This can be used to synchronously render components that have already been loaded,
+ * rather than having to wait for at least one async tick.
+ */
+const syncModuleCache =
+  typeof WeakMap !== 'undefined'
+    ? new WeakMap<() => Promise<any>, React.ComponentType<any>>() // tslint:disable-line:no-any
+    : undefined;
+
+/**
+ * Represents the surface of a module which has a default export.
+ */
+export interface IModuleWithDefault<T> {
+  default: T;
+}
+
+export interface IAsAsyncProps {
+  /**
+   * A placeholder to render while the component loads.
+   */
+  placeholder?: JSX.Element;
+  /**
+   * A callback to be invoked when the component has loaded and rendered.
+   */
+  onLoad?: () => void;
+  /**
+   * A callback to be invoked when the component has failed to load.
+   */
+  onError?: (error: Error) => void;
+}
+
+/**
+ * Produces a component which internally loads the target component before first mount.
+ * The component passes all props through to the loaded component.
+ *
+ * This overload accepts a module with a default export for the component.
+ */
+export function asAsync<TProps, TOwnProps = IAsAsyncProps>(
+  load: () => Promise<{ default: React.ComponentType<TProps> }>,
```

Comment:
assuming the import has a default export is a bad assumption. Instead, could we just have an exportName prop that defaults to `default` and we pull that out of the object?